pysph/tools/ipy_viewer.py

```diff
@@ -634,29 +732,48 @@ def _legend_handler(self, change):
                                 0.82
                             ]
                     )
-                    maxm = np.max(c)
-                    minm = np.min(c)
-                    if (minm == maxm == 0):
-                        boundaries = np.linspace(0, 1, 100)
+
+                    arr = self._normalized_for_colormaps(c)
+                    minm = arr[0]
+                    maxm = arr[1]
+                    absmaxm = arr[2]
+
+                    # Do not change the order of the if-else #
+                    # block below. It is important that the  #
+                    # checks are performed in this order,    #
+                    # otherwise the 'boundaries' array will  #
+                    # not be appropriate.                    #
+
+                    if absmaxm == 0:
+                        boundaries = np.linspace(0, 1, 25)
+                    elif maxm == minm:
+                        # this occurs at initialization for some properties
+                        # like pressure, and stays true throughout for
+                        # others like mass of the particles
+                        boundaries = np.linspace(0, maxm, 25)
                     else:
-                        boundaries = np.linspace(
-                            minm*(1 - np.sign(minm)*0.0001),
-                            maxm*(1 + np.sign(maxm)*0.0001),
-                            100
-                        )
+                        boundaries = np.linspace(minm, maxm, 25)
 
                     self._cbars[array_name] = mpl.colorbar.ColorbarBase(
-                            ax=self._cbar_ax[array_name],
-                            cmap=colormap,
-                            boundaries=boundaries,
+                        ax=self._cbar_ax[array_name],
+                        cmap=colormap,
+                        boundaries=boundaries,
+                        ticks=boundaries,
                     )
                     self._cbars[array_name].set_label(
-                            array_name + " : " +
-                            pa_widgets.scalar.value
+                        array_name + " : " +
+                        pa_widgets.scalar.value
                     )
+        if len(self._cbars.keys()) == 0:
+            self._scatter_ax.set_position(
+                [0, 0, 1, 1]
+            )
+        if change is not None:
+            self.figure.show()
 
     def _save_figure_handler(self, change):
 
+        flag = True
```

Comment:
`flag` is not a good variable name.

Reply:
Changed it to `file_was_saved`, improves readability as well.